Dependancy inversion module initial write up (#4)
Lee-Sutton/clean-code-python: README.md

**[⬆ back to top](#table-of-contents)**

# **Dependency Inversion Principle (DIP)**
The DIP amounts to the following: 
- *Depend upon abstractions. Do not depend upon concrete classes.*

The object of this principle is to help decouple software and render high-level modules independant from low-level module implementation details. In detail, the principle states:
1. High-level modules should not depend on low-level modules. Both should depend on abstractions (e.g. interfaces).
2. Abstractions should not depend on details. Details (concrete implementations) should depend on abstractions.

By dictating that both high-level and low-level objects must depend on the same abstraction, this design principle inverts the way some people may think about object-oriented programming.

> 

Reviewer: I think it would be nice to have another example to illustrate this point. Your example below for Mixins already does a good job of this though so up to you! I personally learn most from reading through examples with code and seeing how these theories and principles can be implemented so in my opinion, the more examples the better 😉

Author: Great point! I might add a visual for this :)

Thinking about the interactions as abstract concepts themselves, allows us to de-couple components while maintaining a lighter and less implementation-dependant interaction schema.

Sources for this module: [Wikipedia](https://en.wikipedia.org/wiki/Dependency_inversion_principle), [Head First Design Patterns](http://ce.sharif.edu/courses/96-97/2/ce484-1/resources/root/Design%20Patterns/Eric%20Freeman,%20Elisabeth%20Freeman,%20Kathy%20Sierra,%20Bert%20Bates-Head%20First%20Design%20Patterns%20-OReilly%20(2008).pdf)

### **Mixins**
Mixins are a special type of multiple inheritance. There are two typical situations that 
require mixins:
1. You want to add in a lot of features and/or methods for a class

> 2. You want to use a particular feature and/or across many different classes

Reviewer 2: I think an important point to note is that a `mixin` should not hold internal state. It should only add functionality via methods or properties

Author: Great point, I might add a tiny section or list of points titled, `what is a mixin?`.

Author: Or maybe, `what is not a mixin?`

> Sometimes we might refer to mixins as being `included` as opposed to `inherited`. Mixins help us abide by the [dependancy inversion principle](#dependency-inversion-principle-dip). Lets walk through the following example where we image we're game makers creating an open world safari adventure game! Let's define some animals:

Reviewer 2: ```suggestion
Sometimes we might refer to mixins as being `included` as opposed to `inherited`. Mixins help us abide by the [dependency inversion principle](#dependency-inversion-principle-dip). Lets walk through the following example where we image we're game makers creating an open world safari adventure game! Let's define some animals:
```

```python 
class Turtle(Hare):
    def swim(self):
        pass
    def crawl(self):
        pass
    def run(self):
        return self.crawl()  
```
Well, okay. That might have fixed it. But does this really make sense? Or is there a *base abstraction* we can pull out here? We will come back to this later. 

We just got an update from marketing that the latest research indicates that gamers really want to play as a Canadian Goose! Well, Canadian Goose swim, run, fly, sleep AND most importantly, poop. So we do the following:

```python 
class CanadianGoose(Hare, Eagle, Turtle):
    def poop(self):
        pass
```
Okay, that was somewhat easy. We have multiple inheritance that seems to cover all of our basses. But lets think about this a little bit. What happens when we ask our Canadian Goose to run? Which class method gets chosen? The same issue goes for when we ask the Canadian Goose to sleep or eat, what would happen? Imagine having to debug this in production! We call this issue the [**Diamond Problem**](https://en.wikipedia.org/wiki/Multiple_inheritance#The_diamond_problem).

So lets go over all our current concerns:
1. This is clearly confusing to any reader. 

> 2. Inherting from concrete classes beging to introduce dependancy webs.

Reviewer 2: ```suggestion
2. Inherting from concrete classes begins to introduce dependency webs.
```

> 3. Debuging can become a nightmare!

Reviewer 2: ```suggestion
3. Debugging can become a nightmare!
```

**Better Design**:

> Imaging we got to go back in time and right our wrongs. What might we have done differently?

Reviewer 2: ```suggestion
Imagine we got to go back in time and right our wrongs. What might we have done differently?
```

1. Instead of having our `Turtle` inherit from a `Hare` we create a *abstract* mixin that they BOTH inherit from.

```python
class AnimalMixin:
    def sleep(self):
        pass
    def eat(self):
        pass

class Hare(AnimalMixin):
    def run(self):
        pass

class Eagle(AnimalMixin):
    def fly(self):
        pass

class Turtle(AnimalMixin):
    def swim(self):
        pass
    def crawl(self):
        pass
```

> 2. Instead of having our Duck inherite from any one class, we can inherit from the AnimalMixin and also create other sets of mixins for the common methods shared across classes. This will allow us to easily identify any potential issues when methods are called and also consolidate code.

Reviewer 2: ```suggestion
2. Instead of having our Duck inherit from any one class, we can inherit from the AnimalMixin and also create other sets of mixins for the common methods shared across classes. This will allow us to easily identify any potential issues when methods are called and also consolidate code.
```
